crates/wasi-http/src/component_impl.rs

```diff
@@ -0,0 +1,422 @@
+use crate::default_outgoing_http::Host;
+pub use crate::r#struct::WasiHttp;
+use crate::streams::Host as StreamsHost;
+use crate::types::Host as TypesHost;
+use crate::types::Scheme;
+use std::str;
+use std::vec::Vec;
+use wasmtime::AsContext;
+use wasmtime::AsContextMut;
+use wasmtime::Caller;
+use wasmtime::Extern;
+use wasmtime::Memory;
+
+const MEMORY: &str = "memory";
+
+#[derive(Debug, thiserror::Error)]
+enum HttpError {
+    #[error("Memory not found")]
+    MemoryNotFound,
+    #[error("Memory access error")]
+    MemoryAccessError(#[from] wasmtime::MemoryAccessError),
+    #[error("Buffer too small")]
+    BufferTooSmall,
+    #[error("UTF-8 error")]
+    Utf8Error(#[from] std::str::Utf8Error),
+}
+
+impl From<HttpError> for u32 {
+    fn from(e: HttpError) -> u32 {
+        match e {
+            HttpError::MemoryNotFound => 2,
+            HttpError::MemoryAccessError(_) => 3,
+            HttpError::BufferTooSmall => 4,
+            _ => panic!("Unsupported"),
+        }
+    }
+}
+
+fn memory_get<T>(caller: &mut Caller<'_, T>) -> Result<Memory, HttpError> {
+    if let Some(Extern::Memory(mem)) = caller.get_export(MEMORY) {
+        Ok(mem)
+    } else {
+        Err(HttpError::MemoryNotFound)
+    }
+}
+
+/// Get a slice of length `len` from `memory`, starting at `offset`.
+/// This will return an `HttpError::BufferTooSmall` if the size of the
+/// requested slice is larger than the memory size.
+fn slice_from_memory(
+    memory: &Memory,
+    mut ctx: impl AsContextMut,
+    offset: u32,
+    len: u32,
+) -> Result<Vec<u8>, HttpError> {
+    let required_memory_size = offset.checked_add(len).ok_or(HttpError::BufferTooSmall)? as usize;
+
+    if required_memory_size > memory.data_size(&mut ctx) {
+        return Err(HttpError::BufferTooSmall);
+    }
+
+    let mut buf = vec![0u8; len as usize];
+    memory.read(&mut ctx, offset as usize, buf.as_mut_slice())?;
+    Ok(buf)
+}
+
+fn u32_from_memory(memory: &Memory, ctx: impl AsContextMut, ptr: u32) -> Result<u32, HttpError> {
+    let slice = slice_from_memory(memory, ctx, ptr, 4)?;
+    let mut dst = [0u8; 4];
+    dst.clone_from_slice(&slice[0..4]);
+    Ok(u32::from_le_bytes(dst))
+}
+
+/// Read a string of byte length `len` from `memory`, starting at `offset`.
+fn string_from_memory(
+    memory: &Memory,
+    ctx: impl AsContextMut,
+    offset: u32,
+    len: u32,
+) -> Result<String, HttpError> {
+    let slice = slice_from_memory(memory, ctx, offset, len)?;
+    Ok(std::str::from_utf8(&slice)?.to_string())
+}
+
+macro_rules! allocate_guest_pointer {
```

Comment:
The compiler doesn't like this function b/c of Rust borrow semantics when I write it as a function:

```
error[E0382]: borrow of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:343:37
    |
330 |         move |mut caller: Caller<'_, T>, stream: u32, len: u64, ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
339 |             let out_ptr = allocate_guest_pointer(caller, body_len);
    |                                                  ------ value moved here
...
343 |             let memory = memory_get(&mut caller).unwrap();
    |                                     ^^^^^^^^^^^ value borrowed here after move

error[E0382]: use of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:390:55
    |
377 |         move |mut caller: Caller<'_, T>, fields: u32, out_ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
382 |             let tuple_ptr = allocate_guest_pointer(caller, (16 * header_len).try_into().unwrap());
    |                                                    ------ value moved here
...
390 |                 let name_ptr = allocate_guest_pointer(caller, name_len);
    |                                                       ^^^^^^ value used here after move

error[E0382]: use of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:391:56
    |
377 |         move |mut caller: Caller<'_, T>, fields: u32, out_ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
390 |                 let name_ptr = allocate_guest_pointer(caller, name_len);
    |                                                       ------ value moved here
391 |                 let value_ptr = allocate_guest_pointer(caller, value_len);
    |                                                        ^^^^^^ value used here after move

error[E0382]: borrow of moved value: `caller`
   --> crates/wasi-http/src/component_impl.rs:393:41
    |
377 |         move |mut caller: Caller<'_, T>, fields: u32, out_ptr: u32| {
    |               ---------- move occurs because `caller` has type `Caller<'_, T>`, which does not implement the `Copy` trait
...
391 |                 let value_ptr = allocate_guest_pointer(caller, value_len);
    |                                                        ------ value moved here
392 |
393 |                 let memory = memory_get(&mut caller).unwrap();
    |                                         ^^^^^^^^^^^ value borrowed here after move

For more information about this error, try `rustc --explain E0382`.
error: could not compile `wasi-http` due to 4 previous errors
```

Admittedly I'm a Rust newb, so if there is a way to resolve this, please let me know. I tried a few things but I couldn't see a way to do it.

Reply:
I figured out how to fix this, so this is now fixed in the PR.